dispatches/unit_models/hydrogen_turbine_unit.py

```diff
@@ -132,6 +132,11 @@ def build(self):
             source=self.stoic_reactor.outlet,
             destination=self.turbine.inlet)
 
+        # Expression for net mechanical work
+        @self.Expression(self.parent_block().time)
+        def work_mechanical(blk, t):
+            return blk.compressor.work_mechanical[t] + blk.turbine.work_mechanical[t]
```

Comment:
So this is actually one of the hydrogen turbine things I wanted to talk about when you asked if I found the turbine to be ever useful in any of the wind+battery+hydrogen runs.

What I found was that this definition of work_mechanical was rarely producing any work because the required compressor work was greater than the produced turbine work. My notes:

With deltaP equal 0, 10 or 24.01 bar, compressor and turbine's `efficiency_isentropic` both 0.99 and stoic_reactor's `conversion` at 0.99, the work done by the turbine is 20% less than the work required by the compressor
    - compressor's work_mechanical doesn't change with deltaP. Even no compression requires in more work than produced by turbine after material conversion and (i presume) heat generation.

So often I had no work at all from this hydrogen_turbine_unit. So I what I did instead was use a compression loss:

```
    compressor_loss = 0.05
    m.fs.h2_turbine.electricity = Expression(m.fs.config.time,
                                             rule=lambda b, t: (-b.turbine.work_mechanical[t] * (1 - compressor_loss)) * 1e-3)
```

I also set the compressor.efficiency_isentropic = 0.99, stoic_reactor.conversion = 0.6, and turbine.efficiency_isentropic=0.6, which gave me an average conversion rate of 'avg_turb_kWh_per_kgH2': 23 which is close to current values IFAIK.

Reply:
@dguittet I never observed a case where the compressor work was higher than the turbine work. I ran the turbine model just now, and I still observe that the compressor work is lower than the turbine work. I also solved the turbine model with deltaP = 10 bar, and observed the same result. I also verified that compressor's mechanical work changes as a function of deltaP. 
```python
from pyomo.environ import ConcreteModel
from idaes.core import FlowsheetBlock

from dispatches.properties.h2_ideal_vap import configuration as h2_ideal_config
from dispatches.properties.hturbine_ideal_vap import configuration as hturbine_config
import dispatches.properties.h2_reaction as h2_reaction_props
from dispatches.unit_models import HydrogenTurbine
from idaes.models.properties.modular_properties.base.generic_property \
    import GenericParameterBlock

from idaes.core.util.model_statistics import degrees_of_freedom
from idaes.core.solvers import get_solver


def build_turbine_model():
    m = ConcreteModel()
    m.fs = FlowsheetBlock(dynamic=False)

    # Load thermodynamic and reaction packages
    m.fs.h2ideal_props = GenericParameterBlock(**h2_ideal_config)
    m.fs.h2turbine_props = GenericParameterBlock(**hturbine_config)
    m.fs.reaction_params = h2_reaction_props.H2ReactionParameterBlock(
        property_package=m.fs.h2turbine_props,
    )

    # Adding H2 turbine model
    m.fs.h2_turbine = HydrogenTurbine(
        property_package=m.fs.h2turbine_props,
        reaction_package=m.fs.reaction_params,
    )

    # Inlet Conditions of the inlet to the compressor.
    m.fs.h2_turbine.compressor.inlet.flow_mol[0].fix(4135.2)
    m.fs.h2_turbine.compressor.inlet.temperature[0].fix(288.15)
    m.fs.h2_turbine.compressor.inlet.pressure[0].fix(101325)

    m.fs.h2_turbine.compressor.inlet.mole_frac_comp[0, "oxygen"].fix(0.188)
    m.fs.h2_turbine.compressor.inlet.mole_frac_comp[0, "argon"].fix(0.003)
    m.fs.h2_turbine.compressor.inlet.mole_frac_comp[0, "nitrogen"].fix(0.702)
    m.fs.h2_turbine.compressor.inlet.mole_frac_comp[0, "water"].fix(0.022)
    m.fs.h2_turbine.compressor.inlet.mole_frac_comp[0, "hydrogen"].fix(0.085)

    m.fs.h2_turbine.compressor.deltaP.fix(2.401e6)
    m.fs.h2_turbine.compressor.efficiency_isentropic.fix(0.86)

    # Specify the Stoichiometric Conversion Rate of hydrogen
    # in the equation shown below
    # H2(g) + 1/2 O2(g) --> H2O(g) + energy
    # Complete Combustion
    m.fs.h2_turbine.stoic_reactor.conversion.fix(0.99)

    # Turbine Parameters
    m.fs.h2_turbine.turbine.deltaP.fix(-2.401e6)
    m.fs.h2_turbine.turbine.efficiency_isentropic.fix(0.89)

    assert degrees_of_freedom(m) == 0

    solver = get_solver()

    # Begin Initialization and solve for the system.
    m.fs.h2_turbine.initialize()

    results = solver.solve(m, tee=True)

    m.fs.h2_turbine.report()


if __name__ == "__main__":
    build_turbine_model()
```

I looked at the turbine implementation in the RE flowsheet, and I noticed that the isentropic efficiency of the compressor and turbine was set to be 0.99 and 0.6, respectively. This is the reason why you are observing compressor work to be higher than the turbine work. The turbine is very inefficient compared to the compressor, so you produce less work for the same deltaP. In general, compressors are less efficient than turbines from what I know. We have been using 0.86 and 0.89 for compressor and turbine respectively. Please change the efficiency values and see if the issue persists. Also, you can increase the stoichiometric conversion from 0.6 to 0.99.

Reply:
Thanks for checking. When I was seeing the compressor work being higher, the efficiencies of the turbine and the compressor and the conv of the stoic_reactor were all equal to 0.99, and I've since lowered it to get the turb conversation ratio down to about 20 kWh / kg H2. Good to know it's working on your end, I'll check again to see if the compressor is behaving as expected.

Regardless, what average conversion rate are you getting for the turbine? This is how I've been calculating it:

```
turbine_elec = [pyo.value(blks[i].fs.h2_turbine.electricity[0]) for i in range(n_time_points)]
turb_kgH2 = [pyo.value(blks[i].fs.h2_turbine.compressor.inlet.flow_mol[0] * blks[i].fs.h2_turbine.compressor.inlet.mole_frac_comp[0, "hydrogen"]) * 3600 / h2_mols_per_kg for i in range(n_time_points)]

...
        "avg_turb_kWh_per_kgH2": np.average(np.array(turbine_elec)/np.array(turb_kgH2)),
 
```

Reply:
Okay, thanks for info on compressor vs turbine efficiencies, I hadn't known that at all. I was able to get turbine work >= compressor work with compressor efficiency = 0.86, stoic reaction = 0.99, and turbine efficiency = 0.89, with deltaP=10 * 1e5. But the overall efficiency seems pretty low: 

```
pyo.value(m.fs.h2_turbine.turbine.work_mechanical[t])*1e-3, pyo.value(m.fs.h2_turbine.compressor.work_mechanical[t])*1e-3
>> (-1169.9374650628813, 1164.094744335233)
```

Reply:
Thank you for checking @dguittet ! Could you tell me what is the composition of the inlet to the compressor? For a deltaP of 10 bar, I got
```
(m.fs.h2_turbine.compressor.work_mechanical[0].value / 1e6, m.fs.h2_turbine.turbine.work_mechanical[0].value / 1e6)
>> (78.02065460139774, -134.0812264447935)
```
So, the net work is 56 MW

Reply:
Ah, I just found that I hadn't set my `purchased_hydrogen_feed` composition so it was just using default values. Now that I've fixed that, the efficiency's doing better. Here is the report output:

```
====================================================================================
Unit : fs.h2_turbine.compressor                                            Time: 0.0
------------------------------------------------------------------------------------
    Unit Performance

    Variables: 

    Key                   : Value      : Units         : Fixed : Bounds
    Isentropic Efficiency :    0.86000 : dimensionless :  True : (None, None)
          Mechanical Work : 1.1637e+06 :          watt : False : (None, None)
          Pressure Change : 1.0000e+06 :        pascal :  True : (None, None)
           Pressure Ratio :     11.000 : dimensionless : False : (None, None)

------------------------------------------------------------------------------------
    Stream Table
                                     Units         Inlet     Outlet  
    Total Molar Flowrate          mole / second     117.60     117.60
    Total Mole Fraction hydrogen  dimensionless   0.085217   0.085217
    Total Mole Fraction nitrogen  dimensionless    0.70196    0.70196
    Total Mole Fraction oxygen    dimensionless    0.18793    0.18793
    Total Mole Fraction water     dimensionless   0.021959   0.021959
    Total Mole Fraction argon     dimensionless  0.0029279  0.0029279
    Temperature                          kelvin     300.00     629.92
    Pressure                             pascal 1.0000e+05 1.1000e+06
====================================================================================

====================================================================================
Unit : fs.h2_turbine.stoic_reactor                                         Time: 0.0
------------------------------------------------------------------------------------
    Unit Performance

    Variables: 

    Key                  : Value  : Units         : Fixed : Bounds
    Reaction Extent [R1] : 4.9607 : mole / second : False : (None, None)

------------------------------------------------------------------------------------
    Stream Table
                                     Units         Inlet     Outlet  
    Total Molar Flowrate          mole / second     117.60     112.64
    Total Mole Fraction hydrogen  dimensionless   0.085217 0.00088970
    Total Mole Fraction nitrogen  dimensionless    0.70196    0.73288
    Total Mole Fraction oxygen    dimensionless    0.18793    0.15217
    Total Mole Fraction water     dimensionless   0.021959    0.11101
    Total Mole Fraction argon     dimensionless  0.0029279  0.0030568
    Temperature                          kelvin     629.92     1297.2
    Pressure                             pascal 1.1000e+06 1.1000e+06
====================================================================================

====================================================================================
Unit : fs.h2_turbine.turbine                                               Time: 0.0
------------------------------------------------------------------------------------
    Unit Performance

    Variables: 

    Key                   : Value       : Units         : Fixed : Bounds
    Isentropic Efficiency :     0.89000 : dimensionless :  True : (None, None)
          Mechanical Work : -1.9547e+06 :          watt : False : (None, None)
          Pressure Change : -1.0000e+06 :        pascal :  True : (None, None)
           Pressure Ratio :    0.090909 : dimensionless : False : (None, None)

------------------------------------------------------------------------------------
    Stream Table
                                     Units         Inlet     Outlet  
    Total Molar Flowrate          mole / second     112.64     112.64
    Total Mole Fraction hydrogen  dimensionless 0.00088970 0.00088970
    Total Mole Fraction nitrogen  dimensionless    0.73288    0.73288
    Total Mole Fraction oxygen    dimensionless    0.15217    0.15217
    Total Mole Fraction water     dimensionless    0.11101    0.11101
    Total Mole Fraction argon     dimensionless  0.0030568  0.0030568
    Temperature                          kelvin     1297.2     778.25
    Pressure                             pascal 1.1000e+06 1.0000e+05
====================================================================================
```

```
pyo.value(m.fs.h2_turbine.turbine.work_mechanical[t])*1e-3, pyo.value(m.fs.h2_turbine.compressor.work_mechanical[t])*1e-3
(-1954.7450392329554, 1163.691024259483)

turb_kw = pyo.value(m.fs.h2_turbine.turbine.work_mechanical[t] + pyo.value(m.fs.h2_turbine.compressor.work_mechanical[t])) * 1e-3
turb_h2kg = pyo.value(m.fs.h2_turbine.compressor.inlet.flow_mol[0] * m.fs.h2_turbine.compressor.inlet.mole_frac_comp[0, "hydrogen"]) * 3600 / h2_mols_per_kg
turb_kw/turb_h2kg
-10.963268365739337
```

If you force the turbine to be used through a reasonable range of power, what's the average kW/kg H2 you get?

Reply:
Sorry, Darice! I missed your earlier comment. Do you mean force it to use in the price-taker problem? The way the turbine model is setup (h2_turbine_work / h2_flowrate) will be constant, provided deltaP and the h2_to_air ratio is fixed. For the conditions I have in the nuclear flowsheet, the turbine produces 12.6 kWh/kg H2.

Reply:
No worries, thanks that seems reasonable if not on the low side (but I'm not too knowledgeable here).